src/Einsum.jl

```diff
@@ -27,336 +28,404 @@ macro einsum_checkinbounds(ex)
-        throw(ArgumentError(
-            string("Threaded @vielsum needs can't assign to a scalar LHS. ",
-                   "Try @einsum instead.")))
+    if threads && length(lhs_indices) == 0
         # this won't actually cause problems, but won't use threads
+        throw(ArgumentError(string(
+            "Threaded @vielsum can't assign to a scalar LHS. ",
+            "Try @einsum instead.")))
     end
 
-    # Copy equation, ex is the Expr we'll build up and return.
-    unquote_offsets!(ex)
+    # Copy equation, expr is the Expr we'll build up and return.
+    unquote_offsets!(expr)
 
     # Next loops to iterate over the destination variables
-    if length(rhs_idx) > 0
+    if length(rhs_indices) > 0
         # There are indices on rhs that do not appear in lhs.
         # We sum over these variables.
 
         if !threads # then use temporaries to write into, as before
 
             # Innermost expression has form s += rhs
             @gensym s
-            ex.args[1] = s
-            ex.head = :(+=)
+            expr.args[1] = s
+            expr.head = :(+=)
 
             # Nest loops to iterate over the summed out variables
-            ex = nest_loops(ex, rhs_idx, rhs_dim, simd, false)
+            expr = nest_loops(expr, rhs_indices, rhs_axis_exprs, simd, false)
 
             # Prepend with s = 0, and append with assignment
             # to the left hand side of the equation.
-            lhs_assignment = Expr(ex_assignment_op, lhs, s)
+            lhs_assignment = Expr(assignment_op, lhs, s)
 
-            ex = quote
+            expr = quote
                 local $s = zero($T)
-                $ex
+                $expr
                 $lhs_assignment
             end
 
         else # we are threading, and thus should write directly to lhs array
 
-            ex.args[1] = lhs
-            ex.head = :(+=)
+            expr.args[1] = lhs
+            expr.head = :(+=)
 
-            ex = nest_loops(ex, rhs_idx, rhs_dim, simd, false)
+            expr = nest_loops(expr, rhs_indices, rhs_axis_exprs, simd, false)
 
-            ex = quote
+            expr = quote
                 $lhs = zero($T)
-                $ex
+                $expr
             end
         end
 
         # Now loop over indices appearing on lhs, if any
-        ex = nest_loops(ex, lhs_idx, lhs_dim, false, threads)
+        expr = nest_loops(expr, lhs_indices, lhs_axis_exprs, false, threads)
     else
         # We do not sum over any indices, only loop over lhs
-        ex.head = ex_assignment_op
-        ex = nest_loops(ex, lhs_idx, lhs_dim, simd, threads)
+        expr.head = assignment_op
+        expr = nest_loops(expr, lhs_indices, lhs_axis_exprs, simd, threads)
     end
 
     if inbounds
-        ex = :(@inbounds $ex)
+        expr = :(@inbounds $expr)
     end
 
     full_expression = quote
-        $ex_create_arrays
+        $type_definition
+        $output_definition
+        $(dimension_checks...)
         let
```

Comment:
One more thought. Does this let block do anything? `quote` already wraps in `begin end` thus local T shouldn't escape.